src/proof/mtree.rs

```diff
@@ -0,0 +1,343 @@
+        write!(f, "{}", hex::encode(self.0))
+    }
+}
+
+impl fmt::Display for BranchElement {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", hex::encode(self.0))
+    }
+}
+
+impl From<[u8; HASH_LENGTH]> for Hash {
+    fn from(b: [u8; HASH_LENGTH]) -> Self {
+        Self(b)
+    }
+}
+
+impl From<[u8; WORD_LENGTH]> for Word {
+    fn from(b: [u8; WORD_LENGTH]) -> Self {
+        Self(b)
+    }
+}
+
+impl From<[u8; BRANCH_ELEMENT_LENGTH]> for BranchElement {
+    fn from(b: [u8; BRANCH_ELEMENT_LENGTH]) -> Self {
+        Self(b)
+    }
+}
+
+impl From<[Hash; 2]> for BranchElement {
+    fn from([h1, h2]: [Hash; 2]) -> Self {
+        let mut b: [u8; BRANCH_ELEMENT_LENGTH] = [0; BRANCH_ELEMENT_LENGTH];
+        b[0..16].copy_from_slice(h1.deref());
+        b[16..32].copy_from_slice(h2.deref());
+        Self(b)
+    }
+}
+
+impl Into<[H256; 4]> for Word {
+    fn into(self) -> [H256; 4] {
+        self.into_h256_array()
+    }
+}
+
+impl Deref for Hash {
+    type Target = [u8; HASH_LENGTH];
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl Deref for Word {
+    type Target = [u8; WORD_LENGTH];
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl Deref for BranchElement {
+    type Target = [u8; BRANCH_ELEMENT_LENGTH];
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+#[derive(Debug, Clone)]
+struct BranchNode {
+    hash: Option<Hash>,
+    left: Option<Rc<Self>>,
+    right: Option<Rc<Self>>,
+    element_on_left: bool,
+}
+
+impl BranchNode {
+    fn accept_left_sibling(&self, hash: Hash) -> Self {
+        Self {
+            hash: None,
+            left: Some(Rc::new(Self {
+                hash: Some(hash),
+                left: None,
+                right: None,
+                element_on_left: false,
+            })),
+            right: Some(Rc::new(self.clone())),
+            element_on_left: false,
+        }
+    }
+
+    fn accept_right_sibling(&self, hash: Hash) -> Self {
+        Self {
+            hash: None,
+            right: Some(Rc::new(Self {
+                hash: Some(hash),
+                left: None,
+                right: None,
+                element_on_left: false,
+            })),
+            left: Some(Rc::new(self.clone())),
+            element_on_left: true,
+        }
+    }
+}
+
+#[derive(Debug)]
+struct BranchTree {
+    raw_data: Word,
+    hashed_data: Hash,
+    root: BranchNode,
+}
+
+#[derive(Debug)]
+struct Node {
+    hash: Hash,
+    count: usize,
+    branches: HashMap<usize, Rc<BranchTree>>,
+}
+
+impl Clone for Node {
+    fn clone(&self) -> Self {
+        Self {
+            hash: self.hash,
+            count: self.count,
+            branches: HashMap::new(),
+        }
+    }
+}
+
+#[derive(Debug)]
+struct NodeElement {
+    next: Option<Box<Self>>,
+    prev: Option<Box<Self>>,
+    value: Node,
+}
+
+#[derive(Debug)]
+struct MerkleTree {
```

Comment:
same here, we will remove it once we get something like https://docs.rs/mrklt/ to work